### PR TITLE
fix(alerts): fix error toast when editing report with saved tab selection

### PR DESCRIPTION
**Pull request by sadpandajoe**
### SUMMARY

**What happens:** Opening the edit modal for a report/alert that has a saved tab selection shows a misleading "There was an error retrieving dashboard tabs" error toast.

**When it happens:** The dashboard tabs API call succeeds, but the response processing in the `.then()` handler crashes when:
1. The selected tab has no scoped native filters — `nativeFilters[anchor]` is `undefined`, so `.map()` throws a TypeError
2. The dashboard has zero native filters — `nativeFilters` is an empty object `{}`, same crash path
3. `all_tabs` is `null` — the `in` operator throws on a non-object operand

The outer `.catch()` swallows the TypeError and shows the tabs error toast, even though the API itself returned valid data.

**Fix:** Add null guards and optional chaining in the tabs `useEffect` handler (`AlertReportModal.tsx`):
- `nativeFilters ?? {}` — prevents `Object.values(undefined)` crash
- `nativeFilters?.[anchor] ?? []` — prevents `.map()` on `undefined` when anchor key is missing
- `allTabs &&` / `!allTabs ||` guards — prevents `in` operator crash on null
- `nativeFilters?.all` — prevents crash when `native_filters` key is omitted entirely
- Replaced 3 `any` type annotations with `NativeFilterObject`

**Tests:** 7 new regression tests covering all crash paths.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A — the error toast no longer appears; no visual changes.

### TESTING INSTRUCTIONS

1. Create a dashboard with tabs and at least one native filter
2. Create a report targeting that dashboard and select a specific tab
3. Open the edit modal for the report
4. Verify no "There was an error retrieving dashboard tabs" error toast appears
5. Verify tab selection and filter dropdowns work correctly
6. Also test with a dashboard that has tabs but zero native filters

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

FEATURE_ALERT_REPORTS=true
FEATURE_ALERT_REPORT_TABS=true
FEATURE_ALERT_REPORTS_FILTER=true

### CODE REVIEW

File: superset-frontend/src/features/alerts/AlertReportModal.test.tsx [hunk 1]
```diff
@@ -105,8 +107,8 @@ const generateMockPayload = (dashboard = true) => {
 const FETCH_DASHBOARD_ENDPOINT = 'glob:*/api/v1/report/1';
 const FETCH_CHART_ENDPOINT = 'glob:*/api/v1/report/2';
 
-fetchMock.get(FETCH_DASHBOARD_ENDPOINT, { result: generateMockPayload(true) });
-fetchMock.get(FETCH_CHART_ENDPOINT, { result: generateMockPayload(false) });
+fetchMock.get(FETCH_DASHBOARD_ENDPOINT, { result: generateMockPayload(true) }, { name: FETCH_DASHBOARD_ENDPOINT });
+fetchMock.get(FETCH_CHART_ENDPOINT, { result: generateMockPayload(false) }, { name: FETCH_CHART_ENDPOINT });
```
[the remark this answers is not in the document]
Comment (sadpandajoe) [reply superset-frontend/src/features/alerts/AlertReportModal.test.tsx]: Thanks for the catch! The code is already formatted in multi-line style — the diff hunk may have been from an earlier revision before the merge commit.

File: superset-frontend/src/features/alerts/AlertReportModal.test.tsx [hunk 2]
```diff
@@ -793,3 +799,437 @@ test('filter reappears in dropdown after clearing with X icon', async () => {
...
+    userEvent.click(screen.getByTestId('contents-panel'));
+    await screen.findByText(/test dashboard/i);
+
+    await waitFor(() => {
+      expect(
+        fetchMock.callHistory.calls().some(c =>
+          c.url.includes('/dashboard/1/tabs'),
+        ),
+      ).toBe(true);
+    });
```
[the remark this answers is not in the document]
Comment (sadpandajoe) [reply superset-frontend/src/features/alerts/AlertReportModal.test.tsx:1015]: Good eye! The `setupAnchorMocks` helper already calls `fetchMock.callHistory.clear()` at the top (line 935), which prevents false positives from prior tests' call history. The diff hunk in the review was truncated before that line.

File: superset-frontend/src/features/alerts/AlertReportModal.tsx [hunk 3]
```diff
@@ -1102,14 +1102,14 @@ const AlertReportModal: FunctionComponent<AlertReportModalProps> = ({
           }
 
           setTabOptions(tabTree);
-          setTabNativeFilters(nativeFilters);
+          setTabNativeFilters(nativeFilters ?? {});
 
-          if (isEditMode && nativeFilters.all) {
+          if (isEditMode && nativeFilters?.all) {
```
Comment (msyavuz) [inline superset-frontend/src/features/alerts/AlertReportModal.tsx:1100]: How was this not an error before if it can be undefined?

File: superset-frontend/src/features/alerts/AlertReportModal.tsx [hunk 4]
```diff
@@ -1121,30 +1121,33 @@ const AlertReportModal: FunctionComponent<AlertReportModalProps> = ({
               const parsedAnchor = JSON.parse(anchor);
               if (!Array.isArray(parsedAnchor)) {
                 // only show filters scoped to anchor
+                const anchorFilters: NativeFilterObject[] =
+                  nativeFilters?.[anchor] ?? [];
```
Comment (msyavuz) [inline superset-frontend/src/features/alerts/AlertReportModal.tsx:1118]: Same thing here, are we missing something if this didn't error before your changes?

File: superset-frontend/src/features/alerts/AlertReportModal.tsx (hunk 3 above)
Comment (sadpandajoe) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:1100]: It *was* erroring — that's the bug this PR fixes. While the backend always returns `native_filters` (at minimum `{}`), when it's an empty dict, accessing `nativeFilters.all` returns `undefined`. That itself is harmless since the `if` condition evaluates falsy. But the related access on line 1125 (`nativeFilters[anchor].map(...)`) throws a TypeError when the anchor key doesn't exist in `native_filters`, because `.map()` is called on `undefined`. That exception bubbles to the `.catch()` handler which shows the misleading "error retrieving dashboard tabs" toast — even though the API request succeeded. The optional chaining guards here are for consistency and defensive typing across all the `nativeFilters` accesses.

File: superset-frontend/src/features/alerts/AlertReportModal.tsx (hunk 4 above)
Comment (sadpandajoe) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:1118]: This is the primary crash site. On master, when a dashboard tab has no scoped native filters, `nativeFilters[anchor]` returns `undefined`. Calling `.map()` on `undefined` throws a TypeError, which the `.catch()` block interprets as a request failure — showing the misleading "error retrieving dashboard tabs" toast to the user. The `?.` + `?? []` guard ensures we gracefully get an empty array instead of crashing.

File: superset-frontend/src/features/alerts/AlertReportModal.tsx (hunk 3 above)
Comment (msyavuz) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:1100]: I meant accessing a property that doesn't exist should throw a typescript error right? We can probably improve the typing at the root here

Comment (sadpandajoe) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:1100]: The root issue was that response.json.result from SupersetClient is typed as any, so nativeFilters inherited any and TypeScript couldn't flag any property access.

Pushed `1e739c89` which adds a DashboardTabsResponse type for the /dashboard/{id}/tabs endpoint. `native_filters` is typed as `Partial<Record<string, NativeFilterObject[]>>` so indexed reads now return NativeFilterObject[] | undefined — the ?. and ?? [] guards are enforced at compile time. Also updated tabNativeFilters state from useState<object> to match, and removed redundant inline type annotations that are now inferred.

File: superset-frontend/src/features/alerts/AlertReportModal.tsx [hunk 5]
```diff
@@ -1446,8 +1450,8 @@ const AlertReportModal: FunctionComponent<AlertReportModalProps> = ({
       return;
 
     // find specific filter tied to the selected filter
-    const filters = Object.values(tabNativeFilters).flat();
-    const filter = filters.filter((f: any) => f.id === nativeFilterId)[0];
+    const filters = Object.values(tabNativeFilters).flatMap(arr => arr ?? []);
+    const filter = filters.filter(f => f.id === nativeFilterId)[0];
```
[the remark this answers is not in the document]
Comment (sadpandajoe) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:1446]: Thanks for the thorough analysis. Both are valid bugs, but they're pre-existing issues in `addNativeFilterOptions` and `onChangeDashboardFilter` — not introduced or modified by this PR. This PR fixes the crash in the tabs `useEffect` response handler. We'll file a follow-up for the stale filter ID handling.

File: superset-frontend/src/features/alerts/AlertReportModal.tsx [hunk 6]
```diff
@@ -750,7 +753,7 @@ const AlertReportModal: FunctionComponent<AlertReportModalProps> = ({
     nativeFilterData.map(nativeFilter => {
       if (!nativeFilter.nativeFilterId) return;
       const filter = nativeFilters.filter(
```
[the remark this answers is not in the document]
Comment (sadpandajoe) [reply superset-frontend/src/features/alerts/AlertReportModal.tsx:747]: Same as above — valid bug, but pre-existing. The `addNativeFilterOptions` function and its unguarded `filter.targets[0]` access existed before this PR. We'll address stale filter ID handling in a follow-up.

File: superset-frontend/src/features/alerts/AlertReportModal.tsx [hunk 7]
```diff
@@ -679,9 +682,9 @@ const AlertReportModal: FunctionComponent<AlertReportModalProps> = ({
   const fetchDashboardFilterValues = async (
     dashboardId: number | string | undefined,
     columnName: string,
-    datasetId: number | string,
+    datasetId: number | string | null,
     vizType = 'filter_select',
-    adhocFilters = [],
+    adhocFilters: any[] = [],
```
Comment (msyavuz) [inline superset-frontend/src/features/alerts/AlertReportModal.tsx:687]: Nit: Is there a better type here?

**Review by msyavuz — APPROVED** [review]
